src/instruct.jl

```diff
@@ -326,19 +337,53 @@ for G in [:Rx, :Ry, :Rz, :CPHASE]
         instruct!(state, m, locs, control_locs, control_bits)
         return state
     end
-
-    @eval function YaoBase.instruct!(
-            state::AbstractVecOrMat{T},
-            g::Val{$(QuoteNode(G))},
-            locs::NTuple{N, Int},
-            theta::Number
-        ) where {T, N}
-        m = rot_mat(T, g, theta)
-        instruct!(state, m, locs)
-        return state
-    end
 end # for
 
+@inline function YaoBase.instruct!(
+        state::AbstractVecOrMat{T},
+        ::Val{:Rx},
+        (loc, )::Tuple{Int},
+        theta::Number
+    ) where {T, N}
+    a, b = cos(theta / 2), -im*sin(theta / 2)
+    instruct_kernel(state, loc, 1 << (loc - 1), 1 << loc, a, b, b, a)
+    return state
+end
+
+function YaoBase.instruct!(
+        state::AbstractVecOrMat{T},
+        ::Val{:Ry},
+        (loc, )::Tuple{Int},
+        theta::Number
+    ) where {T, N}
+    a, b = cos(theta / 2), sin(theta / 2)
+    instruct_kernel(state, loc, 1 << (loc - 1), 1 << loc, a, -b, b, a)
+    return state
+end
+
+function YaoBase.instruct!(
+        state::AbstractVecOrMat{T},
+        ::Val{:Rz},
+        (loc, )::Tuple{Int},
+        theta::Number
+    ) where {T, N}
+    a, d = exp(-im * theta / 2), exp(im * theta / 2)
+    instruct_kernel(state, loc, 1 << (loc - 1), 1 << loc, a, zero(T), zero(T), d)
+    return state
+end
+
+function YaoBase.instruct!(
+    state::AbstractVecOrMat{T},
+    ::Val{:CPHASE},
+    locs::NTuple{N, Int},
+    theta::Number
+) where {T, N}
+    m = rot_mat(T, Val(:CPHASE), theta)
```

Comment:
Now I feel `rot_mat` might be a not intuitive naming. And as we discussed before, it should be in YaoBase. What about using `getmatrix(::Val, params...)` instead?